connectivity/FEATURE_BLE/source/common/ble_trace_helpers.h

```diff
@@ -200,6 +200,375 @@ static inline const char *to_string(privacy_mode_t mode)
     }
 }
 
+static inline const char* to_string(ble::advertising_type_t advType)
```

Comment:
Seems this was just copying the code above.

`static inline` is the worst of all worlds here - it prevents it from using common out-of-line definitions, forcing bloat whether it inlines or not, unless these are used in exactly one file.

You should only ever use `static inline` in headers for super-super-tiny stuff you're confident will be inlined. (And even that can cause problems - it means such functions can't be called from non-static `inline` functions). Only strong reason I've seen for `static inline` is C99/C++ header compatibility, but it seems that non-static `inline` interoperates fine in mixed C99/C++ these days, as long as you remember to provide C99 `extern` definitions in a C file.

You definitely need to take the `static` off all these.

A couple of the smallest functions above with just `if`/`else` could just possibly be an inlining win, but I doubt it. So I'd agree the `inline` is a bit pointless, but not that bad - the compiler can and probably will just choose not to inline.

Main potential benefit of `inline` is that it would permit constant optimisation. I doubt any of these would get passed constants often though.

PS Linkers can deduplicate strings, unless they're placed alongside the code, but I think only ARMC5 likes to do that. In ARMC5 having these non-inline would permit them to do local sharing of `"PUBLIC"`, `"unknown"` and some others, if `--split-sections` was turned off.

Reply:
@kjbracey-arm Thanks for your insights! Just to be clear, are you suggesting that the "code above", i.e. separate from this PR, also suffers from misuse of `static inline` and should be corrected to prevent bloat? From your analysis, it appears as though none of the functions in the header warrant the use of `static inline`, except _perhaps_ those with simple `if/else` control blocks.

Reply:
I'm talking about the new functions and the existing ones, yes.

None of this header code should be `static inline`. Conceivably the `if/else` ones might benefit from simple `inline`, and some of the other little `mbed_trace_array` ones, I doubt any of the others do unless they're ever invoked with a constant. But simple `inline` shouldn't be harmful to code size - main objection would be on grounds of style or compilation time.

Just stripping all the `static` off would satisfy me, but maybe not Paul.

It's possible that the reason for the flood of `static inline` here is actually just a cascade effect from some inner function being `static inline`. Once you've done that, you can't call it from a plain `inline` function. (Because that would then be a common definition calling a translation-unit-local definition).

That's bitten me with CMSIS `__WFE()` type functions. They're `static inline`, which doesn't pose a bloat issue cos they're tiny and always inlined, but you can't then use them in `inline` functions.

Reply:
I think this all just stemmed from some initial assumption that maybe was true for the first one (some function somewhere that was converting a compile time values in a couple of places) and just got out of hand. These are used from multiples places so as long as static is gone I'm fine with it, I'm sure the compiler will ignore inline anyway.

Reply:
I'll move function definitions into source file and remove `static inline` (watch #14107)